rfcs/20200721-extension-types.md

```diff
@@ -0,0 +1,1304 @@
+pros and cons:
+
+* **Pro:** TensorFlow APIs can be used with a type, even if that type was not
+  designed to be used with TensorFlow.  (However, that type must satisfy the
+  requirements of extension types: namely it must be immutable; it must store
+  all dynamic data in tensors; and all static data must be serializable.)
+
+* **Con:** Multiple users might declare different (incompatible) TypeSpecs for
+  the same type, especially for common types.  In most cases, this will result
+  in an error (which may not be very actionable if the incompatible TypeSpecs
+  are registered by libraries that a user is using).  But in some cases, this
+  could result in silently incorrect behavior -- e.g., if one TypeSpec is
+  registered for a type when a model is saved to a SavedModel, but a different
+  TypeSpec is registered for that type when a model is loaded via tf.hub.
+
+### Registry vs Protocol vs Base Class
+
+The current RFC proposes that the API for defining TensorFlow extension types be
+implemented using protocols.  Two alternative APIs that we considered are:
+
+* **Base classes**: Define an abstract class that must be included as a base
+  class for TensorFlow extension types.  We could have a single base class with
+  an abstract `type_spec` property and an abstract `dispatch` method; or we
+  could have separate "mix-in" base classes for each one.  These base classes
+  might or might not have metaclasses.
+
+* **Registration**: Provide functions which can be used to register the
+  `TypeSpec` and the dispatch handler for an extension type.
+
+The following table summarizes some of the pros and cons of each approach:
+
+Feature                                       | Protocol   | Subclass | Registry
+--------------------------------------------- | ---------- | -------- | -------
+Works well with type annotations                    | Y    |    Y     |    N
+Extension type doesn't need to depend on TensorFlow | Y    |    N     |    N
+Extension types can be added non-invasively ①      | N    |    N     |    Y
+Provides a centralized list of all extension types  | N    |    Y ②   |    Y
+
+① As noted above, allowing extension types to be added non-invasively has both
+pros and cons.
+
+② Requires defining a metaclass.
+
+
+### TypeSpec.name
+
+`TensorSpec` is currently the only `TypeSpec` that defines a `name` attribute.
+To my knowledge, the main use for this attribute is to override the default
+argument names for concrete functions when they are called with the flat-
+arguments calling convention.  The question has been raised as to whether all
+`TypeSpecs` should have a `name` attribute.  My initial feeling is that they
+should not, because the core concept for `TypeSpec` is a type specification, and
+type specifications are generally not named.  However, feedback on this question
+is welcome, especially when grounded in concrete use cases.  (Does
+`TensorSpec.name` get used for other things than concrete signature argument
+naming?)
```

Comment:
I also have the feeling that they should not use a `name` attribute. Rather, it seems to me like `name` is not a good solution for the problem it's trying to solve, and that perhaps we should use this RFC as inspiration to find a better one.

Let me show a few examples of why `name` can make things quite confusing for structured data. These examples are about exported signatures in `SavedModel`, how they change before being saved and after being loaded, and how `name` affects the result. I understand that part of the problem here is probably in how `SavedModel` handles structured data and that `name` can have other uses. Still, even if this is not necessarily generalizable, I think it can help as an example.

For this, let's define a simple named tuple named `Pair` and imagine a hypothetical model function called `test` that gets one as input.
```python
Pair = namedtuple('Pair', ('x', 'y'))
````

Let's start with a simple case where we export a signature with no names in its specs. We name our input pair `inputs`.
```python
# Define specs, export, and load.
specs = Pair(tf.TensorSpec([], tf.int32), tf.TensorSpec([], tf.int32))
signatures = {'test': model.test.get_concrete_function(inputs=specs)}
tf.saved_model.save(model, 'test_model', signatures=signatures)
loaded_model = tf.saved_model.load('test_model')

# The inputs structure we're supposed to be exporting.
print(signatures['test'].structured_input_signature)
((Pair(x=TensorSpec(shape=(), dtype=tf.int32, name='inputs/x'), y=TensorSpec(shape=(), dtype=tf.int32, name='inputs/y')),), {})

# The inputs structure from the loaded model.
print(loaded_model.signatures['test'].structured_input_signature)
((), {'inputs_1': TensorSpec(shape=(), dtype=tf.int32, name='inputs_1'), 'inputs': TensorSpec(shape=(), dtype=tf.int32, name='inputs')})
```

The first issue we get is that the use of `name` is inconsistent between signatures in the original and in the loaded model. Note how one is using the argument name `inputs` and the structure fields `x` and `y`, and the other is just using the argument name for all its flattened contents and disambiguating repetitions. This, combined with the fact that position arguments have been converted to keyword ones and that keyword argument dicts have no reliable order, makes recovering structured data trickier and error-prone.

We can try then using the `name` attribute to help sorting out this. Let's give a name to only one of the tensor specs.
```python
specs = Pair(tf.TensorSpec([], tf.int32, name='x'), tf.TensorSpec([], tf.int32))
# [same export, save and load as before] ...

print(signatures['test'].structured_input_signature)
((Pair(x=TensorSpec(shape=(), dtype=tf.int32, name='x'), y=TensorSpec(shape=(), dtype=tf.int32, name='inputs/y')),), {})

print(loaded_model.signatures['test'].structured_input_signature)
((), {'inputs': TensorSpec(shape=(), dtype=tf.int32, name='inputs'), 'x': TensorSpec(shape=(), dtype=tf.int32, name='x')})
```

Now the presence of a `name` is preventing to use the input argument name `inputs` on the first field. However, this is also dropping the structured `inputs/x` in favor of just `x` in the original model. In the loaded one it sort of helps because automatic disambiguation is no longer needed, yet the other non-named spec item gets the argument name. This also means that partially named specs will change the automatically generated disambiguated names of the rest of specs in a nested structure.

Finally, let's show what happens if we have both names.
```python
specs = Pair(tf.TensorSpec([], tf.int32, name='x'), tf.TensorSpec([], tf.int32, name='y'))
# [same export, save and load as before] ...

print(signatures['test'].structured_input_signature)
((Pair(x=TensorSpec(shape=(), dtype=tf.int32, name='x'), y=TensorSpec(shape=(), dtype=tf.int32, name='y')),), {})

print(loaded_model.signatures['test'].structured_input_signature)
((), {'x': TensorSpec(shape=(), dtype=tf.int32, name='x'), 'y': TensorSpec(shape=(), dtype=tf.int32, name='y')})
```

With both names we no longer have disambiguation. However, note that if we were to have a `Pair` nesting two other `Pair` structures, disambiguation would kick in again because `x` and `y` would repeat. This is because `name` is lacking structural information.

So, my point with these examples is just to show some of the cases where the use of the `name` attribute can get things quite messy. Perhaps this is a good time to think about possible better alternatives that could remove the need of such argument.